daemon: Initial VM Support

jrajahalme opened the pull request:
Add support for joining Cilium nodes running in non-k8s mode to a k8s cluster. The main agent option to enable this is `--join-cluster`, which is needed for any of the new functionality to kick in. K8s namespace can be specified with `--k8s-namespace`. Additional node labels can be configured with `--agent-labels`. KV store options must be set to point to the `clustermesh-apiserver` running in the k8s cluster. `--ipv4-range` must be set so that it does not clash with the k8s cluster. Depending on the interfaces available on the VM, `--ipv4-node` may also need to be set so that the IP address the VM is using is routable from the k8s cluster. For example, the following options may be used from the Cilium development VM running in the same host network with the k8s nodes:
```
  --join-cluster
  --k8s-namespace default
  --agent-labels "app=runtime"
  --kvstore etcd --kvstore-opt etcd.address=192.168.36.11:32379
  --ipv4-node 192.168.36.10
  --ipv4-range 10.192.1.0/24
```

There is a new kv-store key at prefix `cilium/state/noderegister/v1/` for agent registrations. Value format is the same as with `cilium/state/nodes/v1`, which also adds a new `nodeidentity` field, which is expected to be set to 0 by the registering agent. `clustermesh-apiserver` is watching this prefix and when agents register, the information passed in the key value is used to create the following k8s resources:
- CiliumIdentity, from the labels in the registration
- CiliumNode, 1:1 from the key value
- CEP, based on the node labels and the allocated identity

`clustermesh-apiserver` replaces the `0` `nodeidentity` with the allocated identity that the registering agent waits for before completing agent bootstrap.

```release-note
Added support for non-k8s nodes to register to a k8s cluster via new option `--join-cluster`
```

jrajahalme committed:
build: Suppress warnings about invalid image names

When pre-pulling images, detect special characters not allowed in
image names to avoid docker warnings about invalid image names. These
were caused by using environment variables for image names. Those
images will not be pre-pulled.

Pre-pulling of docker images used with buildkit builds is a workaround
for a docker bug causing images sometimes not being pulled
properly. Therefore it is possible that when image names are specified
indirectly via an environment variable, and the image is not
pre-pulled, then it is possible that the build will fail due to the
docker bug.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
node: Add NodeIdentity field

Add an optional NodeIdentity field to Node and CiliumNode. This can be
used to pass the Cilium numeric identity of a node in the cluster. If
missing, either NODE or REMOTE_NODE is assumed depending on the Cilium
agent configuration.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

joestringer commented on operator/k8s_cep_gc.go:
```diff
@@ -67,6 +67,7 @@ func enableCiliumEndpointSyncGC(once bool) {
 		// state.
 		watchers.PodsInit(k8s.WatcherCli(), stopCh)
 	}
+	<-k8sCiliumNodesCacheSynced
```
In `onOperatorStartLeading()` we make a similar call, but restricted to `ipamMode` == `ipamOption.IPAMOperator`. Is this OK to unconditionally call here?

jrajahalme replied:
Good question, I'll have to make sure the node watcher is started in all cases.

jrajahalme replied:
Adding a dummy Node Manager to be used when IPAM mode needs no help from the Operator (e.g, `IPAMCRD`).

joestringer commented on examples/crds/ciliumnodes.yaml:
```diff
@@ -163,6 +163,10 @@ spec:
                 type: object
+              nodeidentity:
+                description: NodeIdentity is the Cilium numeric identity allocated for the node, if any.
+                format: int64
```
Does CRD validation provide a way to restrict this further?

In general, we tend to use 24 bits for security identities with the lowest 16 bits representing the in-cluster identity and the 8 bits above representing the cluster. I'm guessing we will ignore the 40 bits above that, so I'm wondering about the consequences. Is it fine because in practice this is populated by the agent?

jrajahalme replied:
Agent itself uses uint32 type, but I went with uint64 for the CRD as there are other types in the CRD using uint64 for the numeric identity (e.g., `AllowedIdentityTuple.Identity`). Given this precedent I'm inclined to keep this as-is.

joestringer commented on pkg/nodediscovery/nodediscovery.go:
```diff
@@ -183,13 +181,26 @@ func (n *NodeDiscovery) StartDiscovery(nodeName string) {
+		}()
+	}
+
+	n.Manager.NodeUpdated(n.LocalNode)
```
The goroutine above that runs `RegisterNode(...)` would previously occur after this `NodeUpdated()` logic is run, but now there is no such guarantee. Is this OK?

jrajahalme replied:
There is no guarantee of when a go routine is scheduled, so this should not change anything.

joestringer replied:
Previously we had something like:

```
NodeUpdated()
go func()  { .... RegisterNode() ... }
```

Means there was a guarantee that `NodeUpdated()` returned before calling `RegisterNode()`, ie the local datapath is plumbed before propagating the node to the kvstore. That said I can't think of a specific case where letting remote nodes know about the local node before locally plumbing it would make a meaningful difference.

joestringer commented on pkg/labels/labels.go:
```diff
@@ -368,6 +368,19 @@ func (l Labels) StringMap() map[string]string {
+			o[v.Source+"."+v.Key] = v.Value
+		}
+	}
+	return o
```
Should we unit-test this?

jrajahalme replied:
Will do after potential other feedback on the treatment of labels here.

jrajahalme replied:
Done.

joestringer reviewed (approved):
LGTM from my codeowners. I will assume other codeowners pay more attention to other areas.

I didn't look so closely at the details of how the rest of the implementation works, just did a brief scan of areas where logic has been shifted around / may run in a different order and asked questions on those points.

aanm commented on operator/k8s_cep_gc.go:
```diff
@@ -90,17 +91,41 @@ func enableCiliumEndpointSyncGC(once bool) {
-								log.WithField(logfields.Object, podObj).
-									Errorf("Saw %T object while expecting *slim_corev1.Pod", podObj)
-								continue
+								continue // No additional checking for non-Pods
```
Please keep the error in this condition and replace this with something like:
```golang
switch pod := podObj.(type) {
case *slim_corev1.Node:
	continue
case *slim_corev1.Pod:
	// In Kubernetes Jobs, Pods can be left in Kubernetes until the Job
	// is deleted. If the Job is never deleted, Cilium will never receive a Pod
	// delete event, causing the IP to be left in the ipcache.
	// For this reason we should delete the ipcache entries whenever the pod
	// status is either PodFailed or PodSucceeded as it means the IP address
	// is no longer in use.
	if k8sUtils.IsPodRunning(pod.Status) {
		continue
	}
default:
	log.WithField(logfields.Object, podObj).
		Errorf("Saw %T object while expecting *slim_corev1.Pod", podObj)
	continue
}
```

joestringer replied:
(nit on the above feedback: `podObj` is already in the structured logging, so we can avoid `Errorf()` in favour of `Error()`.)

jrajahalme replied:
Will do :-)

aanm replied:
@joestringer @jrajahalme the `Errorf` is printing the type of `podObj` (`%T`), not the object itself.

jrajahalme replied:
Done.

aanm commented on bpf/node_config.h:
```diff
@@ -27,6 +27,7 @@ DEFINE_IPV6(HOST_IP, 0xbe, 0xef, 0x0, 0x0, 0x0, 0x0, 0x0, 0x1, 0x0, 0x0, 0xa, 0x
 #define UNMANAGED_ID 3
 #define HEALTH_ID 4
 #define INIT_ID 5
+#define LOCAL_NODE_ID 6
 #define REMOTE_NODE_ID 6
```
Can't this be `#define LOCAL_NODE_ID REMOTE_NODE_ID`?

jrajahalme replied:
It could, but I don't like it. This file is an example node config header that is used for compiling bpf via make. In actual use Cilium Agent will always write this header file based on the agent runtime configuration, and the values can be different. Defining on in the terms of another would be misleading here.

aanm commented on pkg/k8s/apis/cilium.io/v2/types.go:
```diff
@@ -362,6 +362,11 @@ type NodeSpec struct {
+	// NodeIdentity is the Cilium numeric identity allocated for the node, if any.
+	//
+	// +kubebuilder:validation:Optional
+	NodeIdentity uint64 `json:"nodeidentity,omitempty"`
```
@joestringer the `uint64` in the generated CRD came from here. And yes, we can define a validation to accept values from between `0` and `16777215` (2^24)

jrajahalme replied:
Do we have an example of this somewhere?

aanm replied:
cc @christarazi

aanm commented on daemon/cmd/daemon_main.go:
```diff
@@ -456,6 +456,9 @@ func init() {
 
+	flags.String(option.JoinK8sNamespaceName, "", "Name of the Kubernetes namespace join via kvstore registration")
+	option.BindEnv(option.JoinK8sNamespaceName)
+
```
Why not reuse the `option.K8sNamespaceName` flag? That flag is never going to be used in VMs.

jrajahalme replied:
Was doing that, but then needed an enable/disable flag for the agent registration, so decided to go with this. Agent needs to know it is a VM :-)

jrajahalme replied:
Changed to add `--join-cluster` options, to be reused with `--k8s-namespace` option (unless the namespace label should not be set).

aanm commented on pkg/node/store/store.go:
```diff
@@ -32,6 +36,14 @@ var (
 
+	// NodeRegisterStorePrefix is the kvstore prefix of the shared
+	// store for node registration that will be followed by k8s
+	// namespace name depending on agent's --k8s-namespace option
```
Even in the comment you refer to the other flag :smile: it should be `join-k8s-namespace`

jrajahalme replied:
historical reasons, will fix.

jrajahalme replied:
The fix was to remove the latter part of the comment. Namespace is not part of the key path, it is (only) in the node labels.

aanm commented on pkg/node/store/store.go:
```diff
@@ -102,18 +144,61 @@ func (nr *NodeRegistrar) RegisterNode(n *nodeTypes.Node, manager NodeManager) er
+			return err
+		}
+
+		// drain the channel of old updates first
```
Which old updates? We have just registered the `SharedStore`

jrajahalme replied:
the registration of `SharedStore` will dump the keys from the KV store before returning, which will return any old NodeIdentity if it exists. Since the labels on this run can be different (hence leading to a different numeric identity) we must disregard any such identities received before the new registration.

tklauser commented on daemon/cmd/daemon.go:
```diff
@@ -466,6 +468,13 @@ func NewDaemon(ctx context.Context, epMgr *endpointmanager.EndpointManager, dp d
+		agentLabels[k8sConst.PodNamespaceLabel] = option.Config.JoinK8sNamespace
+		agentLabels[k8sConst.PodNameLabel] = nodeTypes.GetName()
+		agentLabels[k8sConst.PolicyLabelCluster] = option.Config.ClusterName
+		// Set configured agent labels to to local node when not running k8s
```
Nit:
```suggestion
		// Set configured agent labels to local node when not running k8s
```

jrajahalme replied:
Thanks!

tklauser commented on daemon/cmd/daemon_main.go:
```diff
@@ -456,6 +456,9 @@ func init() {
 	flags.String(option.K8sNamespaceName, "", "Name of the Kubernetes namespace in which Cilium is deployed in")
 	option.BindEnv(option.K8sNamespaceName)
 
+	flags.String(option.JoinK8sNamespaceName, "", "Name of the Kubernetes namespace join via kvstore registration")
```
Nit:
```suggestion
	flags.String(option.JoinK8sNamespaceName, "", "Name of the Kubernetes namespace to join via kvstore registration")
```

jrajahalme replied:
This will go away with the new boolean option.

tklauser commented on daemon/cmd/daemon.go:
```diff
@@ -583,6 +592,15 @@ func NewDaemon(ctx context.Context, epMgr *endpointmanager.EndpointManager, dp d
 
+	// Must init kvstore before starting node discovery
+	if option.Config.KVStore == "" {
+		log.Info("Skipping kvstore configuration")
```
Should we error out here in case `option.Config.JoinK8sNamespace` is set, i.e. we're running the agent for a VM/node since we cannot work without the KV store config?

jrajahalme replied:
Added option conflict checking elsewhere in `daemon.go`

tklauser commented on Vagrantfile:
```diff
@@ -269,7 +269,7 @@ Vagrant.configure(2) do |config|
     end
     if ENV["NFS"] then
-        config.vm.synced_folder '.', '/home/vagrant/go/src/github.com/cilium/cilium', type: "nfs", nfs_udp: false
+        config.vm.synced_folder '..', '/home/vagrant/go/src/github.com/cilium', type: "nfs", nfs_udp: false
```
>     vagrant: Share parent of cilium/cilium
>
>     Share the parent directory of cilium/cilium. This shares all the
>     cilium repos with one mount, but requires Cilium directory to be named
>     'cilium' (which should usually be the case).

Cannot comment on the commit message, so adding this here :)

We should probably document in the developer docs that the Cilium directory needs to be named `cilium`.

jrajahalme replied:
Docs added.

tklauser commented on operator/cilium_node.go:
```diff
@@ -30,7 +30,15 @@ import (
+var (
+	// CiliumNodeStore contains all CiliumNodes present in k8s.
+	// Warning: The CiliumNodes stored in the cache are not intended to be
+	// used for Update operations in k8s as some of its fields were are not
```
Nit:
```suggestion
	// used for Update operations in k8s as some of its fields are not
```

jrajahalme replied:
Done.

tklauser commented on operator/cilium_node.go:
```diff
@@ -30,7 +30,15 @@ import (
+	// Warning: The CiliumNodes stored in the cache are not intended to be
+	// used for Update operations in k8s as some of its fields were are not
+	// populated.
+	CiliumNodeStore cache.Store
```
Does this need to be exported? It doesn't look like it's used outside of the `operator` package.

```suggestion
	
	ciliumNodeStore cache.Store
```

jrajahalme replied:
Done.

tklauser commented on operator/cilium_node.go:
```diff
@@ -40,7 +48,8 @@ func startSynchronizingCiliumNodes(apiextensionsK8sClient apiextensionsclientset
 	// introducing a slim version of it.
-	_, ciliumNodeInformer := informer.NewInformer(
+	var ciliumNodeInformer cache.Controller
+	CiliumNodeStore, ciliumNodeInformer = informer.NewInformer(
```
Given the previous suggested change:
```suggestion
	ciliumNodeStore, ciliumNodeInformer = informer.NewInformer(
```

jrajahalme replied:
Done.

tklauser commented on operator/k8s_cep_gc.go:
```diff
@@ -90,17 +91,41 @@ func enableCiliumEndpointSyncGC(once bool) {
+								}
+								podChecked = true
+							case "CiliumNode":
+								podObj, exists, err = CiliumNodeStore.GetByKey(owner.Name)
```
Given the previous suggested change regarding `CiliumNodeStore`:
```suggestion
								podObj, exists, err = ciliumNodeStore.GetByKey(owner.Name)
```

jrajahalme replied:
Done.

tklauser commented on operator/k8s_cep_gc.go:
```diff
@@ -90,17 +91,41 @@ func enableCiliumEndpointSyncGC(once bool) {
-						if err != nil {
-							scopedLog.WithError(err).Warn("Unable to get pod from store")
-							continue
+						var podObj interface{}
```
Cannot comment on the entire `DoFunc`, so adding this here:

Maybe move the entire logic to a named `func` in a follow-up commit/PR? The length and the indentation level of this is becoming quite large, so factoring it into a named `func` would improve readability IMO.

jrajahalme replied:
Done.

tklauser commented on pkg/identity/numericidentity.go:
```diff
@@ -107,6 +108,13 @@ const (
+	// localNodeMutex protects localNodeIdentity
+	localNodeMutex lock.Mutex
+	// localNodeIdentity is the endpoint identity allocated for the local node
+	localNodeIdentity = ReservedIdentityRemoteNode
```
Nit: make this an anonymous struct type to keep the mutex together with the value it protects:

```suggestion
+// localNodeIdentity is the endpoint identity allocated for the local node
var localNodeIdentity = struct {
       lock.Mutex
       identity NumericIdentity
}{
       identity: ReservedIdentityRemoteNode,
}
```

jrajahalme replied:
Much nicer, thanks!

tklauser commented on pkg/identity/numericidentity.go:
```diff
@@ -365,6 +373,23 @@ func (id NumericIdentity) Uint32() uint32 {
+func GetLocalNodeID() NumericIdentity {
+	localNodeMutex.Lock()
+	defer localNodeMutex.Unlock()
+	return localNodeIdentity
```
Given the above change:
```suggestion
	localNodeIdentity.Lock()
	defer localNodeIdentity.Unlock()
	return localNodeIdentity.identity
```

jrajahalme replied:
Thx.

tklauser commented on pkg/identity/numericidentity.go:
```diff
@@ -365,6 +373,23 @@ func (id NumericIdentity) Uint32() uint32 {
+func SetLocalNodeID(nodeid uint32) {
+	localNodeMutex.Lock()
+	defer localNodeMutex.Unlock()
+	localNodeIdentity = NumericIdentity(nodeid)
```
Given the above change:
```suggestion
	localNodeMutex.Lock()
	defer localNodeMutex.Unlock()
	localNodeIdentity.identity = NumericIdentity(nodeid)
```

jrajahalme replied:
Done.

tklauser commented on pkg/node/manager/manager.go:
```diff
@@ -334,8 +334,12 @@ func (m *Manager) NodeUpdated(n nodeTypes.Node) {
+			remoteHostIdentity = identity.NumericIdentity(n.NodeIdentity)
+		} else if n.Source != source.Local {
+			remoteHostIdentity = identity.ReservedIdentityRemoteNode
+		}
```
```suggestion
		ni := identity.NumericIdentity(n.NodeIdentity)
		if ni != identity.IdentityUnknown {
			remoteHostIdentity = ni
		} else if n.Source != source.Local {
			remoteHostIdentity = identity.ReservedIdentityRemoteNode
		}
```

jrajahalme replied:
Ok, see the value of casting before comparing, changed :-)

tklauser commented on pkg/node/manager/manager.go:
```diff
@@ -344,7 +348,7 @@ func (m *Manager) NodeUpdated(n nodeTypes.Node) {
 		// encapsulation.
 		if address.Type == addressing.NodeCiliumInternalIP || m.conf.NodeEncryptionEnabled() ||
-			option.Config.EnableHostFirewall {
+			option.Config.EnableHostFirewall || option.Config.JoinK8sNamespace != "" {
```
Just judging from the option name, it might not be obvious that `option.Config.JoinK8sNamespace != ""` means "VM mode". Maybe add a helper function to `option.DaemonConfig` for this check which is more aptly named? This would also be useful in other places where we check the same condition.

jrajahalme replied:
Maybe we need a more aptly named boolean option and then use `--k8s-namespace` in addition? Any suggestions for names? "VM mode" is confusing, IMO, as k8s nodes are also usually VMs, and the same mode can be used to join bare metal machines to Cilium clusters...

tklauser replied:
I just saw that @aanm also mentioned this in https://github.com/cilium/cilium/pull/13309#discussion_r496199006. Agree that a new boolean "VM mode" option and using `--k8s-namespace` would be better in terms of user experience and code readability. And I also agree that having the option called "VM mode" might be confusing given VM is quite an overloaded term. Maybe @tgraf has a suggestion?

jrajahalme replied:
Defining a new option `--join-cluster` and re-using existing `--k8s-namespace` instead.

christarazi commented on pkg/node/types/node.go:
```diff
@@ -171,6 +173,9 @@ type Node struct {
 	Labels map[string]string
+
+	// NodeIdentity is the numeric identity allocated for the node
+	NodeIdentity uint32
```
Why have differing int sizes? Can we just have both be `uint32` or `uint64`? That'll make it easier to reason about because it's confusing to someone at first glace why they're different. Also avoids useless casts between types.

jrajahalme replied:
Historial reasons, I presume. Size changes would be API breaking (on existing APIs). Here I'm just going with the conventions we already have.

christarazi commented on pkg/node/store/store.go:
```diff
@@ -85,14 +92,49 @@ type NodeManager interface {
+	registerStore *store.SharedStore
+}
+
+type RegisterObserver struct {
```
Nit: Godoc

jrajahalme replied:
Added.

christarazi commented on pkg/node/store/store.go:
```diff
@@ -102,18 +144,61 @@ func (nr *NodeRegistrar) RegisterNode(n *nodeTypes.Node, manager NodeManager) er
+			case <-time.After(defaults.NodeInitTimeout / 10):
+				registerStore.Release()
+				nodeStore.Release()
+				return fmt.Errorf("Timed out waiting for node identity")
```
```suggestion
				return fmt.Errorf("timed out waiting for node identity")
```

jrajahalme replied:
Fixed.

christarazi commented on operator/allocator_providers.go:
```diff
@@ -15,9 +15,35 @@
+}
+
+func (nm *nopNodeManager) Resync(context.Context, time.Time) {
+}
```
Nit: I think we can inline these definitions, just so they take up less space :)

```
func (nm *nopNodeManager) Create(resource *v2.CiliumNode) bool { return true }
...
```

jrajahalme replied:
Done, thanks :-)

christarazi reviewed (approved):
LGTM! Non-blocking minor nit.

pchaigno commented on Vagrantfile:
```diff
@@ -278,7 +278,7 @@ Vagrant.configure(2) do |config|
         # nft -f ./contrib/vagrant/nftables.rules
     else
-        config.vm.synced_folder '.', '/home/vagrant/go/src/github.com/cilium/cilium'
+        config.vm.synced_folder '..', '/home/vagrant/go/src/github.com/cilium'
```
Can we condition this change to some environment variable?

Most of my Cilium forks are at `~/ciliumX`, so sharing the parent directory doesn't end well (Vagrant usually stay stuck on the `chown -R vagrant:vagrant` operation). I suspect this isn't so uncommon for folks who do all their development in the dev. VMs. I have one main fork in my gopath and several others in $HOME to be able to run quick tests/manual checks in parallel.

EDIT: Actually, another solution would be to configure a separate synced folder for the clustermesh-apiserver repository.

jrajahalme replied:
There would need to be an option to point to clustermesh-apiserver repo separately anyway. Then there are other repos (e.g., cilium/proxy). I'll add a `SHARE_PARENT` environment variable instead, so if it is set, the parent directory is shared with the directory name (`cilium`) requirement.

jrajahalme replied:
Done, please have another look :-)

pchaigno commented on Documentation/contributing/development/dev_setup.rst:
```diff
@@ -66,8 +66,15 @@ setup for the Vagrantfile in the root of the Cilium tree depends on a number of
+.. note::
+
+   Vagrantfiles assume that the name of your Cilium directory is
+   ``cilium``. This allows mounting of the parent directory to the VMs,
```
For the same reason as highlighted earlier, I have a bunch of Cilium repositories named `cilium`, `cilium2`, etc. Probably a corner case, but maybe we don't even need to enforce the `cilium` name if we setup a separate synced folder for the `clustermesh-apiserver` repository.

pchaigno reviewed (changes requested):
I went through all changes but I'm less familiar with the Go parts (vs. bpf & Vagrantfiles). LGTM except that it breaks my (and I suspect others') current usage of dev. VMs. I think that can be fixed with a separate synced folder for `clustermesh-apiserver`.

---

I also reviewed with 1.10's support for "VM policies" in mind. I initially thought we would need to plug `LOCAL_NODE_ID` in the datapath in place of `HOST_ID` but since you preserve `HOST_ID`, that shouldn't even be necessary. I suspect we're a lot closer to having "VM policies" than we initially thought (:tada:)

One thing that is likely still broken is the resolution to `HOST_ID` for packets from the host. We currently rely on a packet mark for that (to avoid resolving all masqueraded packets as `HOST_ID`), set by an iptables rule, which I'm guessing we don't have here. I'll try to give this PR a try next week to see if I'm missing anything else.

tklauser reviewed (approved):
LGTM for my review comments. No additional comments apart from what Paul and Chris pointed out. I like the new `--join-cluster` option and think it is much more straight-forward.

Tested this again using the instructions from https://github.com/cilium/cilium/pull/13309#issuecomment-700122239 and it worked like a charm.

jrajahalme committed:
bpf: Add support for LOCAL_NODE_ID

Add a bpf macro for LOCAL_NODE_ID that is only used when setting the
source security ID in encap headers for packets originating from the
local node. This defaults to 6, the same value as REMOTE_NODE_ID. Any
other value may be used, if set during Cilium agent bootstrap before
the node_config.h is written.

This allows representing the Cilium node as an endpoint in the
cluster, allowing endpoint selector based network policy enforcement
for traffic originating from Cilium nodes.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
node: Join a k8s cluster via kvstore

Add Cilium agent option `--join-cluster` has the following effects:

- agent registers the node in a special `register` key in the kv store
- agent waits until kv store updates the registration with a numeric
  identifier allocated for the node labels

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
vagrant: Share parent of cilium directory if SHARE_PARENT is set

Share the parent of the cilium directory if SHARE_PARENT is set. This
shares all the cilium repos with one mount, but requires Cilium
directory to be named 'cilium'.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
operator: Make CEP GC aware of the type of the owner

CEP GC used to delete CEPs whenever a Pod resource with the same name
no longer exists. Change this to inspect the OwnerReferences of the
CEP resource, and keep the existing behavior if CEP is owned by a
Pod. If the CEP is owned by a CiliumNode, then check the existence of
that instead.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
test: Lauch runtime VM in the same network as k8s for the same job

Configure cilium-k8s#{$BUILD_NUMBER}-#{$JOB_NAME}-#{$K8S_VERSION}
private network also for the runtime VM so that it has connectivity to
k8s nodes when run at the same time. This is needed for VM support
testing.

Document prevously undocumented options for Vagrantfiles.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

pchaigno reviewed (approved):
LGTM! I checked that the dev. VMs still work and the related documentation changes look great!

Thanks Jarno! The commit breakdown made this PR easier to review than I anticipated :slightly_smiling_face:
